src/coreComponents/common/lifoStorage.hpp

```diff
@@ -0,0 +1,390 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+#ifndef LIFOSTORAGE_HPP
+#define LIFOSTORAGE_HPP
+
+#if defined( GEOSX_USE_CUDA )
+#  include <cuda.h>
+#  include <cuda_runtime.h>
+#endif
+
+#include <chai/ArrayManager.hpp>
+#include <chai/ChaiMacros.hpp>
+#include "common/fixedSizeDeque.hpp"
+#include <deque>
+#include <future>
+#include <string>
+#include <functional>
+#include "LvArray/src/Array.hpp"
+#include "LvArray/src/ArrayView.hpp"
+#include "common/GEOS_RAJA_Interface.hpp"
+
+
+namespace geosx
+{
+/**
+ * This class is used to store in a LIFO way buffers, first on device, then on host, then on disk.
+ */
+template< typename T >
+class lifoStorage
```

Comment:
For memory motion between host and device you'll need to use a `camp::resource` AKA a CUDA stream. The only place you should need to use `std::thread` and whatnot is in moving data between disk and host memory.

Reply:
Ok I'll have a look at it.

Reply:
Yes. But you should only create a single stream per queue. Don't create a new stream for every operation.

Reply:
Indeed

Reply:
Is there a way to ensure my memcpy will use different streams than my kernels in RAJA ?

Reply:
Ok i see i can add ressources to my for loops and make them asynchronous. Else they will use default stream and be synchronous.